source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,172 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_wrap_vector_synonym.h"
+#include <algorithm>
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    uint32_t instruction_id, uint32_t vector_operand1, uint32_t vector_operand2,
+    uint32_t vec_id, uint32_t pos) {
+  message_.set_instruction_id(instruction_id);
+  message_.set_vector_operand1(vector_operand1);
+  message_.set_vector_operand2(vector_operand2);
+  message_.set_vec_id(vec_id);
+  message_.set_scalar_position(pos);
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+  auto vec1 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand1());
+  auto vec2 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand2());
+
+  // |instruction_id| must refer to an existing instruction.
+  if (instruction == nullptr) {
+    return false;
+  }
+
+  std::unordered_set<SpvOp> valid_opcodes{SpvOpIAdd, SpvOpISub, SpvOpIMul,
+                                          SpvOpSDiv, SpvOpUDiv, SpvOpFAdd,
+                                          SpvOpFSub, SpvOpFMul, SpvOpFDiv};
+
+  if (!valid_opcodes.count(instruction->opcode())) {
+    return false;
+  }
+
+  // |vector_operand1| and |vector_operand2| must exist.
+  if (vec1 == nullptr || vec2 == nullptr) {
+    return false;
+  }
+
+  // The 2 vectors must be the same valid vector type.
+  auto vec1_type_id = vec1->type_id();
+  auto vec2_type_id = vec2->type_id();
+  auto vec_type = ir_context->get_type_mgr()->GetType(vec1_type_id);
+
+  if (vec1_type_id != vec2_type_id) {
+    return false;
+  }
+
+  if (vec_type->AsVector() == nullptr) {
+    return false;
+  }
+
+  // |vec_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.vec_id())) {
+    return false;
+  }
+
+  std::vector<uint32_t> vec_ids{message_.vector_operand1(),
+                                message_.vector_operand2(), message_.vec_id()};
+
+  // |vec_id|, |result_id1| and |result_id2| should be disparate ids.
+  if (fuzzerutil::HasDuplicates(vec_ids)) {
+    return false;
+  }
```

Comment:
Why do the vector ids need to be distinct?

Your fuzzer pass will make them initially distinct but in general they don't need to be - e.g. if the scalar operation is:

````
c = 4 + 4
```

then it would be fine to do:

```
v = ivec4(4, 0, 0, 0);
cv = v[0] + v[0]
// c is synonymous with cv
```

Reply:
I just saw all those comments that remained unresolved. For some reason these comments didnt show up in the conversations and only seem to exist in the file changed tab. I will address them today. Thanks! Indeed there is no indeed to make them distinct.

Reply:
Done.